### PR TITLE
replicating a PR from upstream code

### PR DESCRIPTION
**Pull request by willwade**
https://github.com/nateshmbhat/pyttsx3/pull/186

I've literally copy and pasted here - but works well for me :)

### CODE REVIEW

**Comment by dormant-user** [comment]
@willwade Thanks for the PR, I'm not sure what problem this is addressing. I run `py3-tts` on macOS, Windows and Linux
All 3 runs clean, can you narrow down the problem for me.

**Comment by willwade** [comment]
It’s the on word event handling. It’s broken. But this fixes it

eg. From the docs 

```
import pyttsx3
def onStart(name):
    print 'starting', name
def onWord(name, location, length):
    print 'word', name, location, length
def onEnd(name, completed):
    print 'finishing', name, completed
engine = pyttsx3.init()
engine.connect('started-utterance', onStart)
engine.connect('started-word', onWord)
engine.connect('finished-utterance', onEnd)
engine.say('The quick brown fox jumped over the lazy dog.', 'fox')
engine.runAndWait()

```

it was broken in sapi and working in Mac. This fixes the sapi problem. There are a couple of PRs attempting this in the original code. But this is the simplest and works. There is another one but it drifts from fixing other aspects.

File: pyttsx3/drivers/sapi5.py [hunk 1]
```diff
@@ -158,14 +161,18 @@ def __init__(self):
...
         d = self._driver
         if d._speaking:
             d._proxy.notify('finished-utterance', completed=not d._stopping)
         d._speaking = False
         d._stopping = False
         d._proxy.setBusy(False)
+        
+    def _ISpeechVoiceEvents_Word(self, stream_number, stream_position, char, length):
+        self._driver._proxy.notify(
+            'started-word', location=char, length=length)
```
Comment (dormant-user) [inline pyttsx3/drivers/sapi5.py:178]:
```suggestion
            'started-word', location=char, length=length
)
```

File: pyttsx3/drivers/sapi5.py [hunk 2]
```diff
@@ -61,7 +61,10 @@ def say(self, text):
         self._proxy.setBusy(True)
         self._proxy.notify('started-utterance')
         self._speaking = True
-        self._tts.Speak(fromUtf8(toUtf8(text)))
+        # call this async otherwise this blocks the callbacks
+        # see SpeechVoiceSpeakFlags: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms720892%28v%3dvs.85%29
+        # and Speak : https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms723609(v=vs.85)
+        self._tts.Speak(fromUtf8(toUtf8(text)), 1) # -> stream_number as described in the remarks of the documentation
```
Comment (dormant-user) [inline pyttsx3/drivers/sapi5.py:67]:
```suggestion
        self._tts.Speak(fromUtf8(toUtf8(text)), 1)  # -> stream_number as described in the remarks of the documentation
```

File: pyttsx3/drivers/nsss.py [hunk 3]
```diff
@@ -100,8 +100,8 @@ def stop(self):
     @objc.python_method
     def _toVoice(self, attr):
         return Voice(attr.get('VoiceIdentifier'), attr.get('VoiceName'),
-                     [attr.get('VoiceLocaleIdentifier', attr.get('VoiceLanguage'))], attr.get('VoiceGender'),
-                     attr.get('VoiceAge'))
+                 [attr.get('VoiceLocaleIdentifier', attr.get('VoiceLanguage'))], attr.get('VoiceGender'),
+                 attr.get('VoiceAge', None))
```
Comment (dormant-user) [inline pyttsx3/drivers/nsss.py:104]:
```suggestion
                     [attr.get('VoiceLocaleIdentifier', attr.get('VoiceLanguage'))], attr.get('VoiceGender'),
                     attr.get('VoiceAge'))
```

File: pyttsx3/drivers/_espeak.py [hunk 4]
```diff
@@ -80,7 +89,8 @@ def load_windows_epng3():
 
 
 try:
-    load_linux_ep() or load_linux_epng() or load_linux_epng2() or load_windows_epng1() or load_windows_epng2() or load_windows_epng3()
+    load_macos_espeak() or load_linux_epng2() or load_linux_ep() or load_linux_epng() or load_linux_epng2() or load_windows_epng1() or load_windows_epng2() or load_windows_epng3()
+    print(f"Loaded {dll}")
```
Comment (dormant-user) [inline pyttsx3/drivers/_espeak.py:93]:
```suggestion
```

**Review by dormant-user — APPROVED** [review]

**Comment by dormant-user** [comment]
@willwade Thanks for the update. I've approved the changes after looking up related discussions/PRs and docs
I'll merge it once I have these changes tested
Btw, which OS is this driver causing an issue?
`pyttsx3` originally used `sapi5` only for Windows, however it's the `_espeak` module that is updated, and the function says `load_macos_espeak`, so I'm a bit lost here

**Comment by willwade** [comment]
Oh sorry. That's me trying to solve a second problem. That espeak is broken on macOS. It doesn't find the correct .so file but it turns out even if it finds it that library is not working on macOS.  Infact it's broken on windows too. There's quite a few issues discussing this https://github.com/nateshmbhat/pyttsx3/issues/311 

Still my change does at least link to the right so file when it does work. (Nb I'm working on that in espeak)

File: pyttsx3/drivers/sapi5.py [hunk 5]
```diff
@@ -158,14 +161,19 @@ def __init__(self):
...
         if d._speaking:
             d._proxy.notify('finished-utterance', completed=not d._stopping)
         d._speaking = False
         d._stopping = False
         d._proxy.setBusy(False)
+        
+    def _ISpeechVoiceEvents_Word(self, stream_number, stream_position, char, length):
+        self._driver._proxy.notify(
+            'started-word', location=char, length=length
+)
```
Comment (dormant-user) [inline pyttsx3/drivers/sapi5.py:179]:
```suggestion
       )
```

File: pyttsx3/drivers/sapi5.py [hunk 6]
```diff
@@ -158,14 +161,19 @@ def __init__(self):
...
         if d._speaking:
             d._proxy.notify('finished-utterance', completed=not d._stopping)
         d._speaking = False
         d._stopping = False
         d._proxy.setBusy(False)
+        
+    def _ISpeechVoiceEvents_Word(self, stream_number, stream_position, char, length):
+        self._driver._proxy.notify(
+            'started-word', location=char, length=length
+       )
```
Comment (dormant-user) [inline pyttsx3/drivers/sapi5.py:179]:
```suggestion
        self._driver._proxy.notify(
            'started-word', location=char, length=length)
```

File: pyttsx3/drivers/_espeak.py [hunk 7]
```diff
@@ -80,7 +89,7 @@ def load_windows_epng3():
 
 
 try:
-    load_linux_ep() or load_linux_epng() or load_linux_epng2() or load_windows_epng1() or load_windows_epng2() or load_windows_epng3()
+    load_macos_espeak() or load_linux_epng2() or load_linux_ep() or load_linux_epng() or load_linux_epng2() or load_windows_epng1() or load_windows_epng2() or load_windows_epng3()
```
Comment (dormant-user) [inline pyttsx3/drivers/_espeak.py:92]:
```suggestion
    load_macos_espeak() or load_linux_epng2() or load_linux_ep() or load_linux_epng() or load_linux_epng2() or load_windows_epng1() or load_windows_epng2() or load_windows_epng3()
```

**Review by dormant-user — APPROVED** [review]

**Comment by nachoal** [comment]
Hey @willwade where you able to fix the issue of mac not being able to play any sound? I was able to make `espeak-ng` from my M3 mac and fix the file issues with:
- `export DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH`
- `export ESPEAK_DATA_PATH=/usr/local/share/espeak-ng-data`

but now when I run espeak 'text' it doesn't work, looking at the stdout it seems to be corrupted: 
```shell
espeak-ng 'hello world' --stdout
|x=3�@` �nEAe!za^(�q�D�	10�Q	z}n�@Yow@�UUV�|vw~�&Nu� 5CKSXb_ZRMA5)#���������������"$	
�W<A�/q�+L1BLF^>B.�x���"N1�,l��x�P>�Q�#���                                              E7cWLL[0#���.L��	�)
                                          ,as�,g�mgL �a*Vc�ZqmY���0_-6+���>Y�	���Q����1��Q��'�IU0�33�
                                                                                                       QaT/yh\�R*�HZ�ijB^��(N%�Vd�z�W0��GV>�B6`9I0L0�S�a����hv
rOL�a�A�����w�g/M�l���P�Xli.K����j�h�yo>Q?�os{�o	S3�&zT�G;}^j�ye�0!@"s5vO_��a�p4�;
                    Sq���NyY���9�@
(�;T<�(%
       F����Y#�2�(-f{l����IX<l`�Z�,H�11�FVI�#17~��<�X��T7�e��x]
'�X��>7
DJ�_z	�13&@
             yx25DQ�tAw=�,-��|����)J�L	'�u�G%d2/P #!�!L$���-Rg*uGres<����X4�)U�fIh$$�&a��	�cc��J##A������
```

**Pull request closed without merging**